srtcore/api.cpp

```diff
@@ -2387,18 +2381,14 @@ void CUDTUnited::removeSocket(const SRTSOCKET u)
 
 void CUDTUnited::setError(CUDTException* e)
 {
-    delete (CUDTException*)pthread_getspecific(m_TLSError);
-    pthread_setspecific(m_TLSError, e);
+    SetThreadLocalError(e);
 }
 
 CUDTException* CUDTUnited::getError()
```

Comment:
This function should be completely removed, it's a completely useless forwarder for just one function. You can call `GetThreadLocalError` directly here:
```
CUDTException& CUDT::getlasterror()
{
   return *s_UDTUnited.getError();
}
```

Reply:
This function should set an empty error. The pointer returned is then dereferenced, so no null pointer is allowed to be returned here.
Missed that.
```
CUDTException* CUDTUnited::getError()
{
    if (!GetThreadLocalError())
        SetThreadLocalError(new CUDTException);
    return GetThreadLocalError();
}
```

Reply:
Even better to return a reference here right away to avoid mistakes in the future.
This goes way up to the API, so using a pointer in all those functions is probably wrong.

Reply:
Why do you need a lazy initialization provided that you have a global constructor for it?

Reply:
There is a mismatch in logic scope. From the `CThreadError` point of view having `m_TLSError=NULL` is a valid state.
But `CUDTUnited::getError()` can't return NULL because it is then dereferenced.

Placing this logic inside of `CThreadError` to cover the use case of `CUDTUnited::getError()` is not obvious and thus not very good.

Reply:
Therefore you need to preinitialize `m_TLSError = new CUDTException(0, 0, 0)`.

Reply:
The problem here is that a function like `srt_getlasterror` is often called directly after the transmission function, especially in the nonblocking mode. Just a single pointer check puts an extra performance burden on a single call. Whereas this isn't a problem to simply initialize it with in the constructor and have the problem off-head.

Reply:
> Therefore you need to preinitialize `m_TLSError = new CUDTException(0, 0, 0)`.

In the constructor of `CThreadError`?
The constructor will be called only from the main thread. For other threads `m_TLSError` will remain uninitialized.

Reply:
Ah, forgot that in C++03 you can't use `thread_local` :).

OTOH, you could at least do this for C++11, and the actual lazy-init then would be hidden inside the getter.

OTYOH, I wonder how it doesn't crash now that you have a getter that gets the reference directly without lazy-checking.